puts/harness/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+mod bindings {
+    #![allow(non_upper_case_globals)]
+    #![allow(non_camel_case_types)]
+    #![allow(non_snake_case)]
+    #![allow(improper_ctypes)]
+    #![allow(dead_code)]
+    #![allow(clippy::all)]
+    include!(env!("RUST_BINDINGS_FILE"));
+}
+
+pub use bindings::*;
+
+mod init {
+    include!(env!("RUST_PUTS_INIT_FILE"));
+}
+use std::io;
+
+pub use init::*;
+use libc::{c_char, c_void};
+use puffin::error::Error;
+
+pub type FnRegister = extern "C" fn(put: *const C_PUT_TYPE) -> ();
+
+macro_rules! define_extern_c_log {
+    ( $level:ident, $name:ident ) => {
+        unsafe extern "C" fn $name(message: *const c_char) {
+            log::log!(log::Level::$level, "{}", to_string(message));
+        }
+    };
+}
+
+define_extern_c_log!(Error, c_log_error);
+define_extern_c_log!(Warn, c_log_warn);
+define_extern_c_log!(Info, c_log_info);
+define_extern_c_log!(Debug, c_log_debug);
+define_extern_c_log!(Trace, c_log_trace);
+
+#[no_mangle]
+pub static TLSPUFFIN: C_TLSPUFFIN = C_TLSPUFFIN {
+    error: Some(c_log_error),
+    warn: Some(c_log_warn),
+    info: Some(c_log_info),
+    debug: Some(c_log_debug),
+    trace: Some(c_log_trace),
+    make_result: Some(make_result),
+};
+
+pub unsafe fn to_string(ptr: *const c_char) -> String {
+    use std::ffi::CStr;
+
+    if ptr.is_null() {
+        return "".to_owned();
+    }
+
+    CStr::from_ptr(ptr).to_string_lossy().as_ref().to_owned()
+}
+
+use crate::bindings::{
+    RESULT_CODE_RESULT_ERROR_FATAL, RESULT_CODE_RESULT_IO_WOULD_BLOCK, RESULT_CODE_RESULT_OK,
+};
+
+unsafe extern "C" fn make_result(code: RESULT_CODE, description: *const c_char) -> *mut c_void {
+    let reason = to_string(description);
+
+    let result = Box::new(match code {
+        RESULT_CODE_RESULT_OK => Ok(reason),
+        RESULT_CODE_RESULT_IO_WOULD_BLOCK => Err(CError {
+            kind: CErrorKind::IOWouldBlock,
+            reason,
+        }),
+        RESULT_CODE_RESULT_ERROR_FATAL => Err(CError {
+            kind: CErrorKind::Fatal,
+            reason,
+        }),
+        _ => Err(CError {
+            kind: CErrorKind::Error,
+            reason,
+        }),
+    });
+
+    Box::into_raw(result) as *mut _
```

Comment:
Is this guaranteed to be freed?

Reply:
Maybe we should add a note to the `put.h` file

Reply:
The purpose of this function is actually to make sure that we free the memory corresponding to the result. The only purpose of the tuple `(result_code, description)` is to be passed back to Rust, and the Rust code has no choice but to take ownership back if it wants to read it (and thus it will be freed at some point).

If it was allocated in C, it would be difficult to guarantee that this behaves well in every PUT harness implementation. In particular, note that Rust duplicates the `description` string by calling `to_string` so that all memory pointed by the returned pointer is guaranteed to have been allocated through the Rust allocator.

The only case where memory can be lost is when the C harness don't send the resulting pointer back to Rust, but there is not much we can do about that. In the future, I want to discuss all that in a documentation about how to write C PUTs. The specifics in `put.h` are still subject to changes, so I want to wait before making that kind of statement there.